### PR TITLE
bug(tests): mock reads of host's /sys/class/net via get_sys_class_path

### PR DESCRIPTION
**Pull request by blackboxsw**
Create pytest fixture `disable_sysfs_net` to mock cloudinit.net.get_sys_class_path and 
avoid reading /sys/class/net on the host system. This fixture is only applied sparingly to modules that interact with `cloudinit.net.get_sys_class_path` because the cost globally setting  `autouse=True` on this fixture adds about 10 seconds to the test runtime. In modules where we expect most tests to be consuming `get_sys_class_path` we autouse at the module level, otherwise leverage the `disable_sysfs_net` fixture in individual test methods where needed.

This is an answer to GH #4973 where unitests are being run in non-production environments which exhibit duplicate MAC addresses causing many tests to fail.

The plan is it squash merge these separate commits after review.

## Proposed Commit Message
```
bug(tests): mock reads of host's /sys/class/net via get_sys_class_path

Avoid leaking reads to the underlying host's /sys/class/net files.
Some test environments contain virtual network hardware and
configuration such as Calico network devices which provide
duplicated MAC addresses for each device in /sys/class/net. This
results in errors from unittests calling cloudinit.net.get_interfaces.

Provide a disable_sysfs_net` fixture and use it in net-related
modules or functions to avoid unrelated test failures due to
environmental differences.

Intentionally avoid a global `autouse=True` for this fixtures as adds
around 10 seconds to test runs if used for every single unittest.

This should also resolve test platform failures seen in openeuler.

Fixes GH-4973
```

## Test Steps
```
# ensure all tests except for 2 are mocked and prevent access to host's /sys/class/net/
sed -i 's/return SYS_CLASS_NET/raise AssertionError("should not get here")/' cloudinit/net/__init__.py
# expect 2 failures
tox -e py3
```

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Review by TheRealFalcon — COMMENTED** [review]
> the cost globally setting autouse=True on this fixture adds about 10 seconds to the test runtime

A session-scoped fixture only gets applied once and didn't affect runtime at all in my tests. Since we never write anything to `/sys/class/net`, we shouldn't need to worry about each test getting a unique path, correct? Do you foresee any problems with using the same temporary path to be shared across all of the tests? If not, I think this PR could be simplified to something like this:
```diff
diff --git a/tests/unittests/conftest.py b/tests/unittests/conftest.py
index c99f5d860..73b99d758 100644
--- a/tests/unittests/conftest.py
+++ b/tests/unittests/conftest.py
@@ -52,6 +52,16 @@ FS_FUNCS = {
 }
 
 
+@pytest.fixture(scope="session", autouse=True)
+def disable_sysfs_net(tmpdir_factory):
+    """Disables sysfs net directory"""
+    mock_sysfs = f"{tmpdir_factory.mktemp('sysfs')}/"
+    with mock.patch(
+        "cloudinit.net.get_sys_class_path", return_value=mock_sysfs
+    ):
+        yield mock_sysfs
+
+
 @pytest.fixture
 def fake_filesystem(mocker, tmpdir):
     """Mocks fs functions to operate under `tmpdir`"""
diff --git a/tests/unittests/net/test_init.py b/tests/unittests/net/test_init.py
index 60a44186e..c4be1609e 100644
--- a/tests/unittests/net/test_init.py
+++ b/tests/unittests/net/test_init.py
@@ -26,13 +26,13 @@ class TestSysDevPath(CiTestCase):
         """sys_dev_path returns a path under SYS_CLASS_NET for a device."""
         dev = "something"
         path = "attribute"
-        expected = net.SYS_CLASS_NET + dev + "/" + path
+        expected = net.get_sys_class_path() + dev + "/" + path
         self.assertEqual(expected, net.sys_dev_path(dev, path))
 
     def test_sys_dev_path_without_path(self):
         """When path param isn't provided it defaults to empty string."""
         dev = "something"
-        expected = net.SYS_CLASS_NET + dev + "/"
+        expected = net.get_sys_class_path() + dev + "/"
         self.assertEqual(expected, net.sys_dev_path(dev))
 
```

File: tests/unittests/conftest.py [hunk 1]
```diff
@@ -63,6 +63,19 @@ def fake_filesystem(mocker, tmpdir):
     yield str(tmpdir)
 
 
+@pytest.fixture(scope="session", autouse=True)
+def disable_sysfs_net(request, tmpdir_factory):
+    if hasattr(request, "param") and getattr(request, "param") is False:
```
Comment (TheRealFalcon) [inline tests/unittests/conftest.py:68]: Nit: The `hasattr` call isn't needed, right? `getattr` will return `None` if the argument doesn't exist?

**Review by TheRealFalcon — APPROVED** [review]
I left some inline nits, but no code changes needed.

Also...Wooooooooooah! I knew about indirect parametrization, but I didn't know you could parametrize a session-scoped fixture. If we refactor `disable_subp_usage` and `disable_dns_lookup` to work the same, we could potentially save a lot of test setup time.

File: tests/unittests/conftest.py (hunk 1 above)
Comment (blackboxsw) [reply tests/unittests/conftest.py:68]: True story.

**Comment by blackboxsw** [comment]
> I left some inline nits, but no code changes needed.
> 
> Also...Wooooooooooah! I knew about indirect parametrization, but I didn't know you could parametrize a session-scoped fixture. If we refactor `disable_subp_usage` and `disable_dns_lookup` to work the same, we could potentially save a lot of test setup time.

Yep, I was having fun looking over [pytest docs on this this morning](https://docs.pytest.org/en/7.3.x/reference/reference.html#pytest.Metafunc.parametrize).  And agreed I thought the same for disable_subp/dns and dropping some of those `allow` markers from tox.ini

**Comment by holmanb** [comment]
>>    Also...Wooooooooooah! I knew about indirect parametrization, but I didn't know you could parametrize a session-scoped fixture. If we refactor disable_subp_usage and disable_dns_lookup to work the same, we could potentially save a lot of test setup time.
>
> Yep, I was having fun looking over [pytest docs on this this morning](https://docs.pytest.org/en/7.3.x/reference/reference.html#pytest.Metafunc.parametrize). And agreed I thought the same for disable_subp/dns and dropping some of those allow markers from tox.ini

And additionally we can add some other common-sense mocks that would otherwise be "too expensive to consider".

Nice find!

**Pull request merged**